[FLINK-27450][hive] Upgrade default hive version to fix Hive SessionState initialization issue with jdk11

Tartarus0zm opened the pull request:
## What is the purpose of the change

Refer to [HIVE-21584](https://issues.apache.org/jira/browse/HIVE-21584), choose hive version 2.3.9 to support jdk11

## Brief change log

* update hive version to 2.3.9

## Verifying this change

no

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

Tartarus0zm commented:
@wuchong  @luoyuxia  @beyond1920  please take a look if you have time, thanks

MartijnVisser commented:
@Tartarus0zm Hive doesn't support Java 11, see https://issues.apache.org/jira/browse/HIVE-22415

Tartarus0zm commented:
> @Tartarus0zm Hive doesn't support Java 11, see https://issues.apache.org/jira/browse/HIVE-22415

Yes, hive doesn't fully support jdk11 yet.
This PR is to solve the problem that the current HiveParserSessionState cannot be initialized under JDK11.

luoyuxia commented on pom.xml:
```diff
@@ -149,7 +149,7 @@ under the License.
 		-->
 		<minikdc.version>3.2.0</minikdc.version>
-		<hive.version>2.3.4</hive.version>
+		<hive.version>2.3.7</hive.version>
```
As we will also offically provide flink-sql-connector-hive-2.3.9.jar, could you please update it to 2.3.9  and verify it in with jdk 11?

luoyuxia reviewed (commented):
@Tartarus0zm Thanks for contributions. It will enable hive-connector devs to debug/test with JDK 11 in local environment.

MartijnVisser commented:
> This PR is to solve the problem that the current HiveParserSessionState cannot be initialized under JDK11.

Can you explain how that helps if Hive doesn't work under Java 11?

Tartarus0zm commented:
> > This PR is to solve the problem that the current HiveParserSessionState cannot be initialized under JDK11.
> 
> Can you explain how that helps if Hive doesn't work under Java 11?

@MartijnVisser  AppClassLoader is a subclass of URLClassLoader in JDK8, but AppClassLoader is no longer a subclass of URLClassLoader in JDK11.
The low version Hive SessionState code is like this,
```
final ClassLoader currentLoader = Utilities.createUDFClassLoader((URLClassLoader) parentLoader, new String[]{});
```
Therefore, running with jdk11 will report an error, versions after 2.3.6 fixed this problem.
[HIVE-21584]([https://issues.apache.org/jira/browse/HIVE-21584]) fixed it.

Tartarus0zm commented:
@flinkbot run azure

Tartarus0zm commented:
@MartijnVisser  @luoyuxia  CI has passed, please take a look, thanks

luoyuxia commented:
@Tartarus0zm LGTM. But could you please update the pr titile like "[FLINK-27450][hive] Upgrade default hive-exec version to fix Hive SessionState initialization issue  in jdk11"

Tartarus0zm commented:
> @Tartarus0zm LGTM. But could you please update the pr titile like "[[FLINK-27450](https://issues.apache.org/jira/browse/FLINK-27450)][hive] Upgrade default hive-exec version to fix Hive SessionState initialization issue with jdk11"

@luoyuxia  Thank you for your reminder, we have modified the default hive version, not only the hive-exec version, so I changed the title to `[[FLINK-27450](https://issues.apache.org/jira/browse/FLINK-27450)] [hive] Upgrade default hive version to fix Hive SessionState initialization issue with jdk11`

MartijnVisser commented:
@Tartarus0zm Can you please squash commits & rebase? I can merge it afterwards

Tartarus0zm committed:
[FLINK-27450][hive] Upgrade default hive version to fix Hive SessionState initialization issue with jdk11

Tartarus0zm commented:
> @Tartarus0zm Can you please squash commits & rebase? I can merge it afterwards

@MartijnVisser  yes, I has squash commits & rebase. Let's wait for CI to pass. thanks
